File: book-network/src/main/java/com/rpsh/booknetwork/book/Book.java
package com.rpsh.booknetwork.book;


import com.rpsh.booknetwork.common.BaseEntity;
import com.rpsh.booknetwork.feedback.Feedback;
import com.rpsh.booknetwork.history.BookTransactionHistory;
import com.rpsh.booknetwork.user.User;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

@Setter
@Getter
@Entity
@SuperBuilder
@AllArgsConstructor
@NoArgsConstructor

public class Book extends BaseEntity {

    private String title;
    private String authorName;
    private String isbn;
    private String synopsis;
    private String bookCover;
    private boolean archived;
    private boolean shareable;

    @ManyToOne
    @JoinColumn(name = "owner_id")
    private User owner;

    @OneToMany(mappedBy = "book")
    private Set<Feedback> feedbacks = new LinkedHashSet<>();

    @OneToMany(mappedBy = "book")
    private List<BookTransactionHistory> histories;

    @Transient
    public double getRate() {
        if (feedbacks == null || feedbacks.isEmpty()) {
            return 0.0;
        }


        var rate = this.feedbacks.stream()
                .mapToDouble(Feedback::getNote)
                .average()
                .orElse(0.0);

        return Math.round(rate * 10.0) / 10.0;
    }
}

File: book-network-ui/src/app/services/models/book.ts
/* tslint:disable */
/* eslint-disable */
import { BookTransactionHistory } from '../models/book-transaction-history';
import { Feedback } from '../models/feedback';
import { User } from '../models/user';
export interface Book {
  archived?: boolean;
  authorName?: string;
  bookCover?: string;
  createdBy?: number;
  createdDate?: string;
  feedbacks?: Array<Feedback>;
  histories?: Array<BookTransactionHistory>;
  id?: number;
  isbn?: string;
  lastModifiedBy?: number;
  lastModifiedDate?: string;
  owner?: User;
  rate?: number;
  shareable?: boolean;
  synopsis?: string;
  title?: string;
}
